From c9503099d2599daff4cebab8b43ccde666222ba7 Mon Sep 17 00:00:00 2001
From: Russell Keith-Magee <russell@keith-magee.com>
Date: Tue, 6 Dec 2022 09:09:07 +0800
Subject: [PATCH 6/6] Add support for ssl timeout parameters.

---
 src/gbulb/glib_events.py | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/gbulb/glib_events.py b/src/gbulb/glib_events.py
index aba5f8e..832f119 100644
--- a/src/gbulb/glib_events.py
+++ b/src/gbulb/glib_events.py
@@ -317,6 +317,7 @@ class GLibBaseEventLoop(_BaseEventLoop, GLibBaseEventLoopPlatformExt):
         extra=None,
         server=None,
         ssl_handshake_timeout=None,
+        ssl_shutdown_timeout=None,
     ):
         """Create SSL transport."""
         # sslproto._is_sslproto_available was removed from asyncio, starting from Python 3.7.
@@ -329,10 +330,13 @@ class GLibBaseEventLoop(_BaseEventLoop, GLibBaseEventLoopPlatformExt):
                 " or newer (ssl.MemoryBIO) to support "
                 "SSL"
             )
-        # Support for the ssl_handshake_timeout keyword argument was added in Python 3.7.
         extra_protocol_kwargs = {}
+        # Support for the ssl_handshake_timeout keyword argument was added in Python 3.7.
         if sys.version_info[:2] >= (3, 7):
             extra_protocol_kwargs["ssl_handshake_timeout"] = ssl_handshake_timeout
+        # Support for the ssl_shutdown_timeout keyword argument was added in Python 3.11.
+        if sys.version_info[:2] >= (3, 11):
+            extra_protocol_kwargs["ssl_shutdown_timeout"] = ssl_shutdown_timeout
 
         ssl_protocol = sslproto.SSLProtocol(
             self,
@@ -429,6 +433,7 @@ class GLibBaseEventLoop(_BaseEventLoop, GLibBaseEventLoopPlatformExt):
         server=None,
         backlog=100,
         ssl_handshake_timeout=getattr(constants, "SSL_HANDSHAKE_TIMEOUT", 60.0),
+        ssl_shutdown_timeout=getattr(constants, "SSL_SHUTDOWN_TIMEOUT", 60.0),
     ):
         self._transports[sock.fileno()] = server
 
@@ -438,7 +443,6 @@ class GLibBaseEventLoop(_BaseEventLoop, GLibBaseEventLoopPlatformExt):
                     (conn, addr) = f.result()
                     protocol = protocol_factory()
                     if sslcontext is not None:
-                        # FIXME: add ssl_handshake_timeout to this call once 3.7 support is merged in.
                         self._make_ssl_transport(
                             conn,
                             protocol,
@@ -446,6 +450,8 @@ class GLibBaseEventLoop(_BaseEventLoop, GLibBaseEventLoopPlatformExt):
                             server_side=True,
                             extra={"peername": addr},
                             server=server,
+                            ssl_handshake_timeout=ssl_handshake_timeout,
+                            ssl_shutdown_timeout=ssl_shutdown_timeout,
                         )
                     else:
                         self._make_socket_transport(
-- 
2.37.3

